pkg/controller/storagecluster/reconcile.go

```diff
@@ -23,11 +23,17 @@ import (
 // Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
 func (r *ReconcileStorageCluster) Reconcile(request reconcile.Request) (reconcile.Result, error) {
 	reqLogger := r.reqLogger.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+
+	// Trigger initialization job
+	err := r.Initialize(request)
```

Comment:
This should only happen after it's confirmed that a `StorageCluster` resource still exists. It's best to call this in the loop below with the comment "// Add support for additional resources here"

Reply:
There are two places to do initialization:
1. When the operator first starts up, prior to there being any StorageClusters
1. When a new StorageCluster is first detected.

This PR is mostly concerned with the first.

Reply:
Ah. If this is only concerned with the first, then none of this should be called from Reconcile. The `Reconcile` function won't be called until a `StorageCluster` gets created. And when reconciling a particular StorageCluster, which is the primary resource, the code should only be reconciling other resources that are a "secondary resource" to the primary. Often that involves being owned by the primary resource, but that's not always possible.

Is the issue that there could be multiple StorageClusters, but this "initial" data is shared by all of them? Or it is separate and independent from StorageClusters?

Reply:
Separate and independent. And you're right about the reconcile function, I forgot about that. But I think we can live with that, since we want something watching for and reconciling the overall  initialization, and they technically doesn't need to happen until at least one SotrageCluster exists.

Reply:
Gotcha. Then I'll give a harder sell for the approach of making a new CRD, a dedicated "primary resource" that can be reconciled independently and that directly represents the creation and ability to reset these resources. Mixing two different primary resources into one Reconcile function is a recipe for surprising behavior, and it will make it difficult to communicate failure.

For example: suppose something goes wrong while reconciling the ConfigMap, and the code is not able to carry out creation of some resource. Normally you would put information about that failure into the Status of the primary resource. The ConfigMap doesn't even have a Status, so we can't put anything there. If there are multiple StorageCluster resources, we don't want to put status info about the init data onto whichever StorageCluster CR we happen to be reconciling right now. We need a dedicated primary resource that for the init data that is capable of expressing status. That also gives us the benefit of an explicit purpose-made way to declare the desire for the initial data, rather than using a generic configmap.

If you like the below approach, I could get it coded up in a couple days and let you folks focus just on the details of the data that needs to be created. Or I'd be happy to just support the development if @ashishranjan738 wants to give it a go.

## Proposal

A CRD called something like `OCSInitialData` would be created. The Spec would take no arguments (unless you want granular control of which pieces should be created). The Status would reflect whether the data had been created.

Upon startup of the operator, it will ensure that a `OCSInitialData` resource exists. An example of that approach can be seen in the [kni-operator](https://github.com/mhrivnak/kni-operator/blob/a59ac04/cmd/manager/main.go#L142-L163).

### Reconcile

The OCSInitialDataController will be responsible for reconciling that resource and then creating whatever additional resources are required. Once they've been created, the Reconcile function will write to the CR's status that they were created.

Upon further runs of Reconcile, it will see in the Status that the resources had been created and return without taking any action.

... UNLESS the CR is missing, and thus had been deleted. In that case, it will immediately create a new one. Deletion of the CR would enable the end user to cause a "reset" of the initial data back to factory settings.

### CR Example

```
apiVersion: ocs.openshift.io/v1alpha1
kind: OCSInitialData
metadata:
  name: initdata
  namespace: openshift-storage
spec: {}
status:
  created: true
```

### Alternative for Reset

A different pattern for inducing a reset would be to use timestamps. In the Spec a `resetAfter` timestamp, and in the Status a `lastReset` timestamp.

```
if instance.Spec.resetAfter > instance.Spec.lastReset {
  // do whatever it takes to reset the initial data
  // then update Status.lastReset
}
```

To cause a reset, a user only needs to set the `resetAfter` timestamp to the current time.

Personally I like the deletion flow a little better. It's very concrete and simple to understand. But this flow makes more traditional use of the Spec/Status pattern to declaratively request something and then reflect how the request was handled.

Reply:
Okay, but who creates this CR?

Reply:
Best option is likely to have the operator auto-create it on startup, like the example linked to kni-operator. See the second paragraph under **Proposal**.

Reply:
Ah, okay... but then this also means we would need a separate CR for each cluster initialization, and I generally dislike CR proliferation. One ConfigMap to track OCS and cluster initializations seems nicer to keep track of than 1+N CRs. Further, if this is created in the main function, then deleting it doesn't seem like an option unless it is also automatically recreated as part of the deletion callback... or am I missing something?

Reply:
> and I generally dislike CR proliferation

The first time I read this I thought you were suggesting multiple CRDs would be necessary and that CRD proliferation is bad...which I think is a fair concern to have. However, having multiple CRs doesn't sound like a bad thing. Since you would have a controller whose whole job is to keep reconcile n+1 CRs.

> Further, if this is created in the main function, then deleting it doesn't seem like an option unless it is also automatically recreated as part of the deletion callback... or am I missing something?

In the kni-operator example, the main function ensures that the CR exists and then immediately starts the manager that then sits there and waits for events on primary resource it's watching. Since an event is already there (the CR that we just created), the Reconcile functions is called and we can start the kni-operator. This means that a user could delete the kni-operator CR and we wouldn't do anything to recreate it.

There is another option that is more drastic but before going into that I noticed a comment that you had made earlier:

> and they technically doesn't need to happen until at least one SotrageCluster exists.

If this is true then you could simply create the new `OCSInitialData` CRD and associated controller. Then, as part of your reconciliation for the `StorageCluster` resource, you could ensure whatever `OCSInitialData` CRs exist that you want. Leaving the actual work to the `OCSInitialData` controller.

Reply:
Anything you are considering representing inside a ConfigMap, you could instead represent in a single CR. CR proliferation is not necessarily required. As a bonus you'd have more structured data and a standard widely-adopted way to document and validate that structure.

Overcoming delete is a simple two-pronged strategy.

1. On operator startup, ensure it exists.
2. During reconcile, if it gets a "not found" error trying to retrieve the CR, re-create it.

Reply:
The problem here is that we're left with no way to retrigger the initialization tasks for either the operator or the individual StorageClusters. In addition, the same CRD can not be used for both operator and cluster initializations, so that's at least two CRDs in addition to having one CR for the operator and each cluster. This is all much more complicated than just having a single ConfigMap for what I perceive as little gain.

Reply:
> Overcoming delete is a simple two-pronged strategy.
> 
>     1. On operator startup, ensure it exists.
>     2. During reconcile, if it gets a "not found" error trying to retrieve the CR, re-create it.

I assume this means implementing this at a layer above the `Reconcile()` function?

Even so, this still seems like a lot of work for little gain.

Reply:
Cool. It's hard for us to evaluate the details without knowing more about what these initialization tasks are. The best I can do is encourage you that the standard and proven way to represent the ensured existence of something is to make a CRD for it. It's relatively trivial if you're used to building controllers. But if you want to build it out with the configmap, we can always re-evaluate once there's something concrete to look at.

Reply:
> The problem here is that we're left with no way to retrigger the initialization tasks for either the operator or the individual StorageClusters.

Maybe I'm not understanding something. The last comment I made was specifically to address some of these concerns.

> In addition, the same CRD can not be used for both operator and cluster initializations, so that's at least two CRDs in addition to having one CR for the operator and each cluster.

Because there shouldn't be one CRD for to control multiple types of resources and their lifecycle. The argument here is that controller logic should be simplified.

> This is all much more complicated than just having a single ConfigMap for what I perceive as little gain.

My belief is that the truth lies somewhere in the middle in that I bet the ConfigMap approach will be more complicated to do correctly than is being presented here and it's fair to suggest that the approach @mhrivnak and I are advocating is more difficult than our sales pitch.

Having said that, exposing multiple CRDs and having multiple controllers, is the correct way to do this as opposed to the ConfigMap approach. Other projects may have used it but I think it is fair to state that it is not the "Kubernetes" way.

Admittedly, I'm struggling with how https://github.com/openshift/ocs-operator/pull/11/files#diff-b09bb4f016d49aeed647d980bafd4560R23-R41 is obviously another controller masquerading as just a ConfigMap.

What I recommend is:

1. Create the `OCSInitialData` CRD and associated controller.
1. `OCSInitialDataController` does all of the initialization tasks and marks `OCSInitialData.Status.Created = true`
1. `OCSInitialDataController` on reconcile checks `OCSInitialData.Status.Created` and returns immediately if true
1. Add a `func (r *ReconcileStorageCluster) ensureOCSInitialData(...) error` that ensures the `OCSInitialData` Custom Resource(s) exist...not touching them if they do...but if a user deletes them, we have a watch on them and can recreate them to "Reset".
1. Add this `ensureOCSInitialData` function to the range of Reconcile functions.

Reply:
Okay, I think I have an understanding of how this would be maintained. However, I still honestly don't see the benefit of this other than ideological purity. It also seems like a terrible user experience in comparison to having a single known resource type (ConfigMap) with a fixed name to contain this (relatively trivial) state. Rook has gone with this model for a while, as well, so there's also prior art that this works there.

Reply:
> However, I still honestly don't see the benefit of this other than ideological purity.

If the only benefit that we see from this is ideological purity then I owe you an apology. I honestly believe that this approach is more robust, easier to maintain, and easier to reason about than other approaches.